source/row_selection_table_model.ts

```diff
@@ -1,227 +1,265 @@
+   * @param event - The event describing the keyboard press.
+   */
+  public onKeyUp(event: KeyboardEvent): void {
     const keyCode = event.keyCode;
     if(keyCode === 38) {
       this.isUpDown = false;
       if(this.state === 4) {
-        this.s0();
+        return this.s0();
       } else if(this.state === 7) {
-        this.s0();
+        return this.s0();
       }
     } else if(keyCode === 40) {
       this.isDownDown = false;
       if(this.state === 4) {
-        this.s0();
+        return this.s0();
       } else if(this.state === 7) {
-        this.s0();
+        return this.s0();
       }
     } else if(keyCode === 16) {
       this.isShiftDown = false;
       if(this.state === 2) {
-        this.s2();
+        return this.s2();
       }
     } else if(keyCode === 17) {
       this.isCtrlDown = false;
       if(this.state === 4) {
-        this.s0();
+        return this.s0();
       }
     }
   }
 
-  private C0() {
+  /** Marks the beginning of a transaction. In cases where a transaction is
+   *  already being processed, then the sub-transaction gets consolidated into
+   *  the parent transaction.
+   */
+  private beginTransaction(): void {
+    if(this.transactionCount === 0) {
+      this.operations = [];
+    }
+    ++this.transactionCount;
+  }
+
+  /** Ends a transaction. */
+  private endTransaction(): void {
+    --this.transactionCount;
+    if(this.transactionCount === 0) {
+      this.dispatcher.dispatch(this.operations);
+    }
+  }
+
+  private c0() {
     return this.isShiftDown &&
       (this.isMouseDown || this.isDownDown || this.isUpDown);
   }
-  private C1() {
+
+  private c1() {
     return this.isCtrlDown && this.isMouseDown;
   }
-  private C2() {
+
+  private c2() {
     return !this.isShiftDown && !this.isCtrlDown && this.isMouseDown;
   }
-  private C3() {
+
+  private c3() {
     return !this.isShiftDown && !this.isCtrlDown &&
       (this.isDownDown || this.isUpDown);
   }
-  private C4() {
+
+  private c4() {
     return !this.isMouseDown && !this.isShiftDown;
   }
 
   private s0() {
     this.state = 0;
     if(!this.isMouseDown && !this.isDownDown && !this.isUpDown) {
-      this.currentRow = this.hilightedRow;
+      this.currentRow = this.highlightedRow;
     }
-    if(this.C0() ) {
-      this.s1();
-    } else if(this.C1()) {
-      this.s3();
-    } else if(this.C2()) {
-      this.s5();
-    } else if(this.C3()) {
-      this.s7();
+    if(this.c0()) {
+      return this.s1();
+    } else if(this.c1()) {
+      return this.s3();
+    } else if(this.c2()) {
+      return this.s5();
+    } else if(this.c3()) {
+      return this.s7();
     }
   }
 
   private s1() {
     this.state = 1;
-    this.clearTrueRows();
-    this.previousRow = this.hilightedRow;
+    this.clearRows();
+    this.previousRow = this.highlightedRow;
     this.isAdding = true;
-    this.s2();
+    return this.s2();
   }
 
-  private s2() {
+  private s2(): any {
```

Comment:
Remove the ```any```

Reply:
That might be a problem.The compiler complains that:

's0' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.

Reply:
Okay then TypeScript is unable to infer types of recursive functions. In that case instead of doing ```return this.sXXX()``` you will need to do

```
this.sXXX();
return;
```